### PR TITLE
Added resampling to freerdp_dsp_encode.

### PR DESCRIPTION
**Pull request by akallabeth**
Readds sound resampling for non FFMPEG encoding path

### CODE REVIEW

**Comment by mfleisz** [comment]
@akallabeth Sounds way better now but I can still hear some clicks/distortion ... but not sure what is causing this ...

**Comment by mfleisz** [comment]
@llyzs Can you also have a look at this PR please?

**Comment by llyzs** [comment]
Different number of channels (nChannels) also need to be compared and resampled, please see my above comment.

**Comment by llyzs** [comment]
I fixed some obvious errors, please try: @mfleisz 

```diff --git a/channels/rdpsnd/server/rdpsnd_main.c b/channels/rdpsnd/server/rdpsnd_main.c
index 4be0b3078..2fbda6f06 100644
--- a/channels/rdpsnd/server/rdpsnd_main.c
+++ b/channels/rdpsnd/server/rdpsnd_main.c
@@ -501,7 +501,7 @@ static UINT rdpsnd_server_send_wave2_pdu(RdpsndServerContext* context,
        Stream_Write_UINT16(s, context->selected_client_format); /* wFormatNo */
        Stream_Write_UINT8(s, context->block_no); /* cBlockNo */
        Stream_Seek(s, 3); /* bPad */
-       Stream_Write_UINT16(s, wTimestamp); /* dwAudioTimeStamp */
+       Stream_Write_UINT32(s, wTimestamp); /* dwAudioTimeStamp */
        src = context->priv->out_buffer;
        length = context->priv->out_pending_frames * context->priv->src_bytes_per_frame;

@@ -558,7 +558,7 @@ static UINT rdpsnd_server_send_samples(RdpsndServerContext* context,
        UINT error = CHANNEL_RC_OK;
        EnterCriticalSection(&context->priv->lock);

-       if (context->selected_client_format < 0)
+       if (context->selected_client_format == (UINT16)-1)
        {
                /* It's possible while format negotiation has not been done */
                WLog_WARN(TAG, "Drop samples because client format has not been negotiated.");
@@ -636,7 +636,7 @@ static UINT rdpsnd_server_close(RdpsndServerContext* context)

        if (context->priv->out_pending_frames > 0)
        {
-               if (context->selected_client_format < 0)
+               if (context->selected_client_format == (UINT16)-1)
                {
                        WLog_ERR(TAG, "Pending audio frame exists while no format selected.");
                        error = ERROR_INVALID_DATA;
@@ -652,7 +652,7 @@ static UINT rdpsnd_server_close(RdpsndServerContext* context)
        if (error)
                return error;

-       context->selected_client_format = -1;
+       context->selected_client_format = (UINT16)-1;
        Stream_Write_UINT8(s, SNDC_CLOSE);
        Stream_Write_UINT8(s, 0);
        Stream_Seek_UINT16(s);
@@ -807,7 +807,7 @@ RdpsndServerContext* rdpsnd_server_context_new(HANDLE vcm)
        context->vcm = vcm;
        context->Start = rdpsnd_server_start;
        context->Stop = rdpsnd_server_stop;
-       context->selected_client_format = -1;
+       context->selected_client_format = (UINT16)-1;
        context->Initialize = rdpsnd_server_initialize;
        context->SelectFormat = rdpsnd_server_select_format;
        context->SendSamples = rdpsnd_server_send_samples;
diff --git a/libfreerdp/codec/dsp.c b/libfreerdp/codec/dsp.c
index 22b2d3423..8e44062af 100644
--- a/libfreerdp/codec/dsp.c
+++ b/libfreerdp/codec/dsp.c
@@ -1044,7 +1044,8 @@ BOOL freerdp_dsp_encode(FREERDP_DSP_CONTEXT* context, const AUDIO_FORMAT* srcFor
        if (!context || !context->encoder || !srcFormat || !data || !out)
                return FALSE;

-       if (srcFormat->nSamplesPerSec != context->format.nSamplesPerSec)
+       if (srcFormat->nSamplesPerSec != context->format.nSamplesPerSec ||
+           srcFormat->nChannels != context->format.nChannels)
        {
                if (!freerdp_dsp_resample(context, data, length, srcFormat))
                        return FALSE;

```

**Comment by akallabeth** [comment]
@llyzs Integrated most of the fixes you found, but it looks like the `freerdp_dsp_resample` method is really broken as it was (no proper sample rate conversion, channels ignored, bit depth ignored)
Will take a bit to fix that.

File: libfreerdp/codec/dsp.c [hunk 1]
```diff
@@ -104,46 +109,95 @@ struct _FREERDP_DSP_CONTEXT
...
+	size_t sframes, rframes;
+	size_t rsize;
+	size_t i, j;
+	size_t sbytes, rbytes;
+	size_t srcBytesPerFrame, dstBytesPerFrame;
+	size_t srcChannels, dstChannels;
+	AUDIO_FORMAT format;
+
+	if (!context || !src || !srcFormat || !data || !length)
+		return FALSE;
```
Comment (hardening) [inline libfreerdp/codec/dsp.c]: static function should not self protect themselves against invalid arguments.

Comment (akallabeth) [reply libfreerdp/codec/dsp.c]: @hardening This function is due for a rewrite anyhow ;)
Resampling is quite broken and the sound will be distorted (only exception: `44100Hz` to `22050Hz` or similar special cases)

**Comment by mfleisz** [comment]
@akallabeth Can this be flagged as non-WIP as we should do a separate PR for fixing the re-sampler?

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by mfleisz** [comment]
The project fails to compile with SOXR support because the soxr.h include file cannot be found. Looks like cmake does not add the proper include path when compiling dsp.c

**Comment by akallabeth** [comment]
@mfleisz fixed

File: libfreerdp/codec/dsp.c [hunk 2]
```diff
@@ -1032,7 +1044,14 @@ BOOL freerdp_dsp_encode(FREERDP_DSP_CONTEXT* context, const AUDIO_FORMAT* srcFor
 	if (!context || !context->encoder || !srcFormat || !data || !out)
 		return FALSE;
 
-	// TODO: Resample
+	if (srcFormat->nSamplesPerSec != context->format.nSamplesPerSec)
+	{
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/codec/dsp.c]: @llyzs This is true, but up/downmixing was never supported and will be handled in a follow up pull.

[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/codec/dsp.c]: @llyzs for reference, tracking that in #5519

**Pull request merged**